server/repository/src/tests.rs

```diff
@@ -1181,4 +1181,129 @@ mod repository_test {
+            let connection = manager_a.connection().unwrap();
+            let result: Result<(), TransactionError<RepositoryError>> = connection
+                .transaction_sync(|con| {
+                    println!("A: transaction started");
+                    let repo = ItemRowRepository::new(con);
+                    let _ = repo.find_one_by_id("tx_deadlock_id")?;
+                    println!("A: read");
+                    println!("A: Sleeping for 1000ms");
+                    let start_dt = SystemTime::now();
+                    for _ in 0..2 {
+                        //2*500ms
+                        std::thread::sleep(core::time::Duration::from_millis(500));
+                        println!("A: slept for 500ms");
+                    }
+                    let transaction_duration = SystemTime::now()
+                        .duration_since(start_dt)
+                        .expect("Time went backwards");
+                    println!("Slept for {:?}", transaction_duration);
+                    println!("A: write");
+                    let _ = repo.upsert_one(&inline_init(|i: &mut ItemRow| {
+                        i.id = "tx_deadlock_id2".to_string();
+                        i.name = "name_a".to_string();
+                    }))?;
+                    println!("A: written");
+                    Ok(())
+                });
+            result
+        });
+
+        let process_b = tokio::spawn(async move {
+            //Wait for process a to get a transaction started
+            let connection = connection_manager.connection().unwrap();
+            // let _ = a_finished_reading_receiver.await;
+            println!("B: Ready to start transaction");
+            // println!("Starting transaction in blocking thread...");
+            let result: Result<(), TransactionError<RepositoryError>> = connection
+                .transaction_sync(|con| {
+                    println!("B: transaction acquired");
+                    let repo = ItemRowRepository::new(&con);
+                    let _ = repo.find_one_by_id("tx_deadlock_id")?;
+                    println!("B: read");
+                    let _ = repo.upsert_one(&inline_init(|i: &mut ItemRow| {
+                        i.id = "tx_deadlock_id".to_string();
+                        i.name = "name_b".to_string();
+                    }))?;
+                    println!("B: write 1");
+
+                    let _ = repo.upsert_one(&inline_init(|i: &mut ItemRow| {
+                        i.id = "tx_deadlock_id".to_string();
+                        i.name = "name_b_2".to_string();
+                    }))?;
+                    println!("B: write 2");
+                    Ok(())
+                });
+            println!("B: Returning {:?}", result);
+            result
+        });
+
+        let a = process_a.await.unwrap();
+        let b = process_b.await.unwrap();
+        a.unwrap();
+        b.unwrap();
+    }
```

Comment:
@andreievg Can you clarify what you mean with asserting database lock? Do you mean create a test case to prove that without busy timeout set, we get an immediate error? Feels like the refactoring work is bigger than the benefits here?

Reply:
Added a comment to explain how to trigger the failing case, which I think is probably enough?

Reply:
> Feels like the refactoring work is bigger than the benefits here?

Sure, on closure look it is, if it was just sqlite connection then perhaps I would have pushed to try and get the fail test assertion. Looks like you've added at least some asserts (i think my comment was just about that, was just weird seeing tests with no asserts (although unwraps would have triggered a failure anyway)

Btw `assert!("name_a" == tx_deadlock_item2.name);` is better written as `assert_eq!("name_a", tx_deadlock_item2.name);`  just because assert failure will print values of both sided